File: leetcode/arrays-hashing/longest_consecutive_sequence.ts
function longestConsecutive(nums: number[]): number {
  const numSet = new Set(nums);
  let longest = 0;

  for (const n of numSet) {
    // Only start counting from the beginning of a sequence
    if (!numSet.has(n - 1)) {
      let length = 1;
      while (numSet.has(n + length)) {
        length++;
      }
      longest = Math.max(longest, length);
    }
  }

  return longest;
}

// Try it out — press Run
console.log(longestConsecutive([100, 4, 200, 1, 3, 2])); // 4
console.log(longestConsecutive([0, 3, 7, 2, 5, 8, 4, 6, 0, 1])); // 9
console.log(longestConsecutive([])); // 0

export {};

File: leetcode/arrays-hashing/longest_consecutive_sequence.py
from typing import List

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        num_set = set(nums)
        longest = 0

        for n in num_set:
            # Only start counting from the beginning of a sequence
            if n - 1 not in num_set:
                length = 1
                while n + length in num_set:
                    length += 1
                longest = max(longest, length)

        return longest

# Try it out — press Run
sol = Solution()
print(sol.longestConsecutive([100, 4, 200, 1, 3, 2]))  # 4
print(sol.longestConsecutive([0, 3, 7, 2, 5, 8, 4, 6, 0, 1]))  # 9
print(sol.longestConsecutive([]))  # 0
